### PR TITLE
MAINT: stats.dirichlet: fix interface inconsistency

### LINKED ISSUE #6006: Dirichlet doesn't accept its own random variates as input to pdf
```

In [1]: import scipy.stats as ss

In [2]: d = ss.dirichlet([1,3,4])

In [3]: d.pdf(d.rvs())
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-a6f3762e83cf> in <module>()
----> 1 d.pdf(d.rvs())

/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/scipy/stats/_multivariate.py in pdf(self, x)
   1365
   1366     def pdf(self, x):
-> 1367         return self._dist.pdf(x, self.alpha)
   1368
   1369     def mean(self):

/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/scipy/stats/_multivariate.py in pdf(self, x, alpha)
   1260         """
   1261         alpha = _dirichlet_check_parameters(alpha)
-> 1262         x = _dirichlet_check_input(alpha, x)
   1263
   1264         out = np.exp(self._logpdf(x, alpha))

/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/scipy/stats/_multivariate.py in _dirichlet_check_input(alpha, x)
   1079                          "of entries as, or one entry fewer than, "
   1080                          "parameter vector 'a', but alpha.shape = %s "
-> 1081                          "and x.shape = %s." % (alpha.shape, x.shape))
   1082
   1083     if x.shape[0] != alpha.shape[0]:

ValueError: Vector 'x' must have either the same number of entries as, or one entry fewer than, parameter vector 'a', but alpha.shape = (3,) and x.shape = (1, 3).

In [4]: n = ss.norm()

In [5]: n.pdf(n.rvs())
Out[5]: 0.16964393208168185
```

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-6006
gh-4984

#### What does this implement/fix?
The `dirichlet` distribution interface is inconsistent with other multivariate distributions and even itself: the `pdf` method expects the transpose of what the `rvs` method produces. This PR introduces `multivariate_beta`, which is the same distribution without this inconsistency.

The plan is to deprecate `dirichlet` in favor of `multivariate_beta`. (@h-vetinari can you help with this, either by making a PR against my branch or a follow-up to this one? I can review it.) After the deprecation cycle, we can make `dirichlet` an alias for `multivariate_beta` if desired.

#### Additional information
There are many other things that could be improved about the distribution and its documentation (e.g. `pdf` input `x` can only be 2D). This PR does not fix all of them, but it does fix a defect reported as early as gh-4984. Let's get this messy stuff out of the way, then those other things can be cleaned up in future PRs.

The three commits are pretty clean. I'd suggest reviewing them separately and in order.

LMK if I should address gh-6474 the same sort of way... maybe `inv_wishart`?

### CODE REVIEW

File: scipy/stats/_multivariate.py
Comment (mdhaber) [inline scipy/stats/_multivariate.py:1593]: Copy-paste with minimal changes (`dirichlet` -> `multivariate_beta` as needed)

File: scipy/stats/tests/test_multivariate.py
Comment (mdhaber) [inline scipy/stats/tests/test_multivariate.py:642]: Will have two subclasses - one with `dist = dirichlet`, the other with `dist = multivariate_beta`.

File: scipy/stats/_multivariate.py
Comment (mdhaber) [inline scipy/stats/_multivariate.py:1233]: `dirichlet` and `multivariate_beta` both have a new `_check_input` method. 
`dirichlet` does `x = np.asarray(x)` then calls `_dirichlet_check_input(alpha, x)`.
`multivariate_beta` does `x = np.moveaxis(x, -1, 0)` then calls  `_dirichlet_check_input(alpha, x)`.
That is the only thing that's different between the two distributions, other than documentation.

File: scipy/stats/_multivariate.py
Comment (mdhaber) [inline scipy/stats/_multivariate.py:1702]: Probably could have just done transpose. I didn't realize that `dirichlet` was only written for 2D `x`.

**Review by mdhaber — COMMENTED** [review]
Some comments to facilitate review.

**Comment by h-vetinari** [comment]
> @h-vetinari can you help with this, either by making a PR against my branch or a follow-up to this one?

Hey, sorry I haven't gotten around to this yet. I'm in the middle of preparing a move, and am on low availability until ~middle of May. I'll try to take a look when I can, but no promises, unfortunately... 🙈

File: scipy/stats/_multivariate.py
Comment (tirthasheshpatel) [inline scipy/stats/_multivariate.py:1300]:
```suggestion
other distributions, methods `pdf` and `logpdf` expect the transpose of the
```

File: scipy/stats/_multivariate.py
Comment (tirthasheshpatel) [inline scipy/stats/_multivariate.py:1301]:
```suggestion
input `x`. Please use `multivariate_beta`, which corrects this inconsistency.
```

File: scipy/stats/_multivariate.py
Comment (tirthasheshpatel) [inline scipy/stats/_multivariate.py:1308]: Should we change this to

```suggestion
    r"""A Dirichlet random variable (deprecated, use `multivariate_beta` instead).
```

?

File: scipy/stats/_multivariate.py
Comment (tirthasheshpatel) [inline scipy/stats/_multivariate.py:1656]:
```suggestion
    If the quantiles don't lie within the simplex, a `ValueError` is raised.
```

I think we link Python docs so adding a backtick around Python keywords should link its documentation.

File: scipy/stats/tests/test_multivariate.py
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_multivariate.py:851]: wow, didn't know this existed :)

File: scipy/stats/tests/test_multivariate.py
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_multivariate.py:874]: are these `assert`s really necessary? We can just test if the call to pdf succeeds.

**Review by tirthasheshpatel — COMMENTED** [review]
Only small nitpicks. Otherwise LGTM.

One question: Unlike univariate distributions, we are not constrained by the infrastructure to add new parameters to the pdf and logpdf methods. Instead of deprecating the whole distribution, we can alternatively add a new keyword to the pdf and logpdf method (e.g. `transpose`) and default it to True. We can then emit a deprecation warning when it is true and default it to false in 1.11.0. To me, this sounds simpler than adding a new distribution. Hove you considered doing this?

**Comment by mdhaber** [comment]
I don't remember! I think I got carried away after gh-15889 (which also might interest you) and maybe didn't stop to think. Good idea.

On the other hand, we'd have the unfortunate choice of getting stuck with a `transpose` keyword or backward incompatibility (default to `transpose=None` but give the option of `transpose=True` for temporary relief).

Would you like to submit a PR for that, and if it's merged, we'd close this?

File: scipy/stats/_multivariate.py
Comment (mdhaber) [reply scipy/stats/_multivariate.py:1656]: Sure. I don't think I was fixing the documentation here. There are still a few issues in multivariate documentation : )

**Comment by mdhaber** [comment]
Actually this needs an email to the mailing list. I'll send one with both options and let the default be to merge this one if there are no comments in favor of the other? Might as well since the work is done.

File: scipy/stats/tests/test_multivariate.py
Comment (mdhaber) [reply scipy/stats/tests/test_multivariate.py:874]: I sort-of remember now - I wanted to test both the frozen and non-frozen versions. Maybe I figured that since I was calling both, I might as well check that they're identical, too?

File: scipy/stats/_multivariate.py
Comment (mdhaber) [inline scipy/stats/_multivariate.py:1656]:
```suggestion
    If the quantiles don't lie within the simplex, a ``ValueError`` is raised.
```
How about this, then?

Comment (tirthasheshpatel) [reply scipy/stats/_multivariate.py:1656]: Looks good!

File: scipy/stats/__init__.py
Comment (mdhaber) [inline scipy/stats/__init__.py:164]:
```suggestion
   dirichlet              -- Dirichlet (deprecated, see `multivariate_beta`)
```

File: scipy/stats/_multivariate.py
Comment (mdhaber) [inline scipy/stats/_multivariate.py:1298]:
```suggestion
"""  # noqa
```

**Comment by mdhaber** [comment]
OK, if that commit resolved the PEP8 issues, is this ready to merge after giving some time for people to respond to the email?

File: scipy/stats/tests/test_multivariate.py
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_multivariate.py:874]: Oh, ok. It does cover the frozen distribution code. It's OK to keep them then.

**Review by tirthasheshpatel — APPROVED** [review]
LGTM! Added just one small comment. We can wait till Sunday for feedback on the mail. If there is none, this should be good to go in!

File: scipy/stats/tests/test_multivariate.py
Comment (mdhaber) [reply scipy/stats/tests/test_multivariate.py:874]: Actually I think that's another part of what I was doing - testing vectorization against a loop.

**Comment by mdhaber** [comment]
+1 -0 from the mailing list, @tirthasheshpatel. Would you like to follow up with `multivariate_gamma`/`multivariate_invgamma` and I review, or would you be willing to review another few like these?

**Review by tylerjereddy — APPROVED** [review]
Between the PR review and mailing list comments, I'll take this as +2 core developers in favor and CI is passing apart from a timeout.

I scanned through the diff and it looked well-done from a mechanical (non-stats-expert) point of view, but I'm mostly leaning on the current code review/feedback.

**Comment by tylerjereddy** [comment]
thanks

**Comment by tirthasheshpatel** [comment]
Thanks @mdhaber!

> Would you like to follow up with `multivariate_gamma`/`multivariate_invgamma` and I review, or would you be willing to review another few like these?

Sure, I can propose a PR for `multivariate_gamma` and `multivariate_invgamma`.

**Comment by oscarbenjamin** [comment]
I just picked this up in SymPy CI (https://github.com/sympy/sympy/pull/23513). I went to follow the instruction to use `multivariate_beta` but it doesn't exist in older SciPy versions:
```
  File "/opt/hostedtoolcache/Python/3.11.0-beta.1/x64/lib/python3.11/site-packages/numpy/lib/utils.py", line 95, in newfunc
    warnings.warn(depdoc, DeprecationWarning, stacklevel=2)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
DeprecationWarning:         `dirichlet.rvs` is deprecated, use `multivariate_beta.rvs` instead!
          # noqa
        `dirichlet` is deprecated due to an interface inconsistency: compared to
        other distributions, methods `pdf` and `logpdf` expect the transpose of the
        input `x`. Please use `multivariate_beta`, which corrects this inconsistency.
        In SciPy 1.11.0, this deprecation warning will be removed and `dirichlet` will
        become an alias for `multivariate_beta`.
```
What should be the expected code to use if wanting to suppress the warning while supporting multiple SciPy versions?

Something like this?
```python
from scipy import stats

dirichlet = getattr(stats, 'multivariate_beta', None) or stats.dirichlet
```
And then in the future when we don't need to worry about old SciPy versions should we change the code again to just use `multivariate_beta` unconditionally? I think it's better if downstream code only needs to be updated once in response to something like this.

Generally I think that if there does not already exist an undeprecated alternative API then it is better to have a period of "soft deprecation" before emitting warnings or making any breaking change. By soft deprecation I mean something like:

1. Add the new better API.
2. Point to the new API in the docs and warn (in the docs) against using the old API.
3. *Maybe* emit PendingDeprecationWarning so it can be seen in CI but not by end users.

It's not completely clear to me what is different between `dirichlet` and `multivariate_beta` or what will be changed about `dirichlet` in future. Is it only the `pdf` and `logpdf` methods that are different? Or is it the `rvs` method that was changed to make it consistent with the others?

If we were using those methods then presumably we would need to change something else in the code rather than simply replacing `dirichlet` with `multivariate_beta`. I think it's good to make that clear like:
```
The multivariate_beta class is a dropin replacement for dirichlet
except that the pdf and logpdf methods use a different convention.
If you use these methods like this

    from scipy.stats import dirichlet
    p = dirichlet.pdf(x)
    logp = dirichlet.logpdf(x)

then you should change that code to

    from scipy.stats import multivariate_beta
    p = multivariate_beta.pdf(np.transpose(x))
    logp = multivariate_beta.logpdf(np.transpose(x))

All other methods of multivariate_beta and dirichlet are identical
and any code using them does not need to be changed.
```
It's important when writing something like this to consider that the person (e.g. maintainer of large codebase) trying to fix the downstream code might know very little about the API and what it is used for and really needs clear instructions for what to do.

On the other hand SymPy doesn't use the `pdf` or `logpdf` methods so if that's all that has been changed then maybe there is no reason to give a warning if only the `rvs` method is being used (assuming that hasn't been changed and won't be in future).

**Comment by mdhaber** [comment]
> What should be the expected code to use if wanting to suppress the warning while supporting multiple SciPy versions?

Yes, that, or `try` the `multivariate_beta` import and fall back to `dirichlet`.

> PendingDeprecationWarning

That sounds like a project level decision. I don't think I've seen that used before. 

> if that's all that has been changed then maybe there is no reason to give a warning if only the rvs method is being used

For now, only `pdf` and `logpdf` has changed, and the only difference is that the expected input is the transpose of what it was before. So we could change this emit only on use of `pdf` and `logpdf`.

There is a separate change that may be made to RVS soon to fix an old bug.

**Comment by mdhaber** [comment]
@tirthasheshpatel Thanks for [offering](https://github.com/scipy/scipy/pull/16042#issuecomment-1136725841) to add `multivariate_gamma` and `multivariate_invgamma`. In the end, I think that in gh-16277 we'll add `multivariate_beta` without yet deprecating `dirichlet`, so I'd suggest adding those two classes without deprecating `wishart` and `invwishart`.

You are also slated to address gh-7689, which would fix the shape of the `rvs` output for all of these distributions. It would be nice if we could at least fix this in `multivariate_beta`, `multivariate_gamma`, and `multivariate_invgamma` before branching. (Otherwise we should probably hold up the release of `multivariate_beta` so we don't have to change it later.) Let me know if this is too much with the little time we have!

**Comment by tirthasheshpatel** [comment]
> It would be nice if we could at least fix this in `multivariate_beta`, `multivariate_gamma`, and `multivariate_invgamma` before branching. (Otherwise we should probably hold up the release of `multivariate_beta` so we don't have to change it later.) Let me know if this is too much with the little time we have!

I will submit a PR for `multivariate_beta` and `multivariate_invgamma` ASAP (should I do that in one PR or propose a separate one for each?). I will also submit one fixing the `size=1` issue and see if we can get it merged before 1.9 branches.

**Comment by mdhaber** [comment]
Let's do separate ones. Thank you!

**Comment by mdhaber** [comment]
@tirthasheshpatel I would prioritize fixing the `squeeze` issues of `multivariate_beta` in all methods. If we can't do that before branch, I'd suggest holding the addition of `multivariate_beta` until we can (i.e. reverting).

**Pull request merged**
